### ISSUE TITLE
Custom dispatchers with varied action types

### ISSUE THREAD

**Issue by zigster64** [karlseguin/http.zig#64]
# situation # 

I have an app with a dispatch method that defines an action type that takes *App

so far so good … I can create handlers that take *App as the 1st param

Now I have added an extra custom dispatcher that takes a custom context - let’s say, a struct that wraps *App and a Session object

When I add that route : 

router.getC(“/account”, accountHandler, protectedRouteDispatcher)

that’s now a problem .. it’s using a comptime lookup of the vanilla dispatch method to get the type of the action param .. which is different to the type of the action param for the protectedRouteDispatcher

throws a compile error on the router.getC call, saying accountHandler wants an action type of *App, rather than the wrapped ctx

# solution # 

just needs a bit more comptime reflection to get those action types aligned I think

way past bedtime here, will have a go at solving it tomorrow 

.. I think it’s a simple fix, need to sleep on it. I have a hacky oauth2 repo posted that demonstrates the error in the meantime. 

# half solution # 

is an easy fix if I make all my dispatchers take the exact same action type :)

however, I think it’s a good plan to enable multiple different action types, as that will open the door to allowing 3rd party middleware plugins to play nicer together

for example - if someone published an oauth/session management plugin that defined its own dispatcher and action type.

**Comment by zigster64** [comment]
https://github.com/zigster64/oauth.zig/blob/f485c00f16a3321fea801f58bee69bb45886ed15/src/app.zig#L161

My custom dispatcher that has a different action type to the default

**Comment by karlseguin** [comment]
I've thought about this before. I agree that it _should_ work that way. I disagree that it's simple. However, my Zig-fu is better now versus when I wrote this, so I'm more than happy to look into it again and try to come up with a solution.

The issue is that the router is a `std.StringHashMap(func(*App, *Request, *Response) !void)` (1).  I'm not sure how you can make the router be a lookup for different actions.

My first two thoughts are: 1. Expose the action as an *anyopaque, pass that to the app's dispatch method, and let it do the casting. 2. Use a union for the route context. 2 might already work as is, so instead of a `httpz.Action(*SessionCtx)`, you'd have a `httpz.Action(Action)` where `Action` is your own union. 

(1) - it's really a trie, but a simpler view helps focus on the issue.

**Comment by zigster64** [comment]
Hmmm .. tricky. I see what you mean. Not that simple at all.

couple of considerations here

1. if the goal is to enable loading hypothetical 3rd party middleware/dispatchers, then anyopaque might be the way to go. Not sure yet. If using a union, then the app author would need to be aware of the action types needed in the 3rd party code, and add that to its own union. (I think)
2. I’m thinking the benefit of having both a simple dispatcher with just *app, and a complex dispatcher with a struct that wraps app + other stuff is that the simple case doesn’t suffer any extra overhead where it’s not needed. That supposed extra overhead might not be that bad once the compiler optimizes everything anyway.

Will have a play with my hacky app and see if I can solve my issue with using a union .. get it working, then have a think about how to present a really easy API to do this - preferably with having the bare minimum amount of ptr casting :)

Another way out of this might be to have multiple routers attached to the server - each router having its own consistent action type, so the problem of creating a collection of different fn pointers goes away.

**Comment by karlseguin** [comment]
That's would be nice, if it worked (although, if that worked, we could probably infer the value from the parameters of the dispatcher).

Maybe I'm being blinded by the current implementation (which I'm more than happy to change). The problem isn't knowing the type, the issue is storing the type for later use when executing.  What does `addRouter` + `router.get` do, that lets us get an `*SessionCtx` on `GET /users` but an `*App` on /data/list` ? Zig doesn't let you work on `Type` at runtime, i.e. you can store `Type` in a field.

How do you store a `*SessionCtx` alongside an `*App` (or their function pointer equivalent if you're storing the action). *anyopaque, union or 1 router per type, and then check every router for a match.

I'm trying to approach this outside of http.zig, just to come up with a possible design. I'm starting with the "problem":

```zig
const std = @import("std");

pub fn main() !void {
  var gpa = std.heap.GeneralPurposeAllocator(.{}){};
  const allocator = gpa.allocator();

  var x = std.StringHashMap(???).init(allocator);
  defer x.deinit();

  try x.put("/1", f1);
  try x.put("/2", f2);
}

fn f1(_: u32) void {}
fn f2(_: bool) void {}
```

**Comment by zigster64** [comment]
Yep.
I’m suggesting using multiple hashmaps - one for each type

addRouter would create a new hashmap internally … each having its own action type and dispatcher as well

when matching a request- check through multiple hashmaps until it finds a hit

that’s the theory, will have a play and see if that works

If it works the way I think it will, then the getC / putC variants are not strictly needed anymore either

**Comment by zigster64** [comment]
hmm.

TL;DR - Changing the httpz design is harder than it looks.  Changing the users app to use a consistent ActionType is less painful than in looks.
 
Implementing "multilple dynamic hash maps" is harder than it looks for sure.  You are right - it comes back to needing to store the type in some way, and that can only really be done at compile time via a comptime func that returns a type.  

It's almost like you need to declare an array of StringHashMap(anytype) and let the compiler generate a collection of StringHashMap(T) based on whats used.   ugh !

I finished up my ProofOfConcept app, with a view to using that as a test mule to attempt addressing the above issue

What I found there is that most of the "Simple routes that only need *App" ... ended up needing the full Session context anyway, for various reasons which were not apparent at the start. Once the Session ctx appears in the app, you start running into the situation where things that dont use the context need to at least be aware of it and act accordingly.

So Ive ended up with everything using the same ActionType anyway :(

I think the general pattern for anything more complicated than *App is always going to be a struct that wraps *App, and includes other session related bits.   

Im trying to think up some cases now where as a user of the lib - I see some nice 3rd party httpz middleware that I want to use that "conforms to some common interface" .. and I can slot that into my already custom app without a lot of drama.  Same as way Go http middleware can be slotted in.

So with that in mind, should keep this concept on the back burner for a while and keep thinking about it for a while.

**Comment by karlseguin** [comment]
I've always ended up in the same situation as you. I have an "Env" which wraps an `*App* and an `?*User` and stuff like that.

It does feel like most middleware could be composed/called from the custom dispatch. That was always the point of the dispatch getting the `action` to call...you can wrap it, or not call it.

Stilll, there's now a [middlewares](https://github.com/karlseguin/http.zig/tree/middlewares) branch. It's incomplete, but I think the hard part is done.

There's an example of what a [middleware looks like](https://github.com/karlseguin/http.zig/blob/middlewares/example/middleware/Logger.zig). And [simple.zig](https://github.com/karlseguin/http.zig/blob/middlewares/example/simple.zig#L25) shows how a user might use it for a [group of] route.

Two main things still to do:
1. I plan on making the server's allocator and arena available to the middleware in case they have any allocations to make, and also to support a calling their `deinit` if they define one.

2. I'll probably add a `StringHashMap(*anyopaque)` to the request object. This would let the middleware store arbitrary data:

```zig
pub fn execute(self: *const Auth, req: *httpz.Request, res: *httpz.Response, executor: anytype) !void {
  const user = loadUser(...);
  req.middlewares("auth", user);
}
```

And then provide a function (or many functions) for the action to call:

```zig
fn addProduct(app: *App, req: *httpz.Request, res: *httpz.Response) !void {
  const user = middleware.User.getUser(req);
}
```


I might also get rid of the C variants of the router, and just make it default:

```
get(path, action) -> get(path, action, config)
```

I think I originally thought the "C"onfig variants would rarely be used, but by adding middleware, they'll get used more and it's just weird now.

**Comment by karlseguin** [comment]
Middlewares are now in master.
